src/ImageManipulation.php

```diff
@@ -709,6 +711,33 @@ public function ThumbnailURL($width, $height)
         return $this->getIcon();
     }
 
+    /**
+     * Convert the file to another format if there's a registered converter that can handle it.
+     *
+     * @param string $toFormat The file extension you want to convert to - e.g. "webp".
+     * @param bool $returnNullOnFailure If true and the conversion fails, `null` is returned instead.
+     * This is useful if calling this method from a template, for example, where you don't want to be throwing
+     * exceptions for end users.
+     * @throws FileConversionException If the conversion fails and $returnNullOnFailure is false.
+     */
+    public function convert(string $toFormat, bool $returnNullOnFailure = false): ?AssetContainer
```

Comment:
We intentionally don't expose `$options` here because this is intended to be primarily used from templates (e.g. `$MyFile.convert('webp')`). Passing an array of options in templates is I think unsupported but at best would be complicated to do. If developers really need that level of control of their file conversions, they can use the `FileConversionManager` in their controller or model.

Reply:
If we need to pass in options from templates in projects we pass them as json_encoded array or object, and access it using `func_get_args` as it can't have a default value.

Which brings me to the second param here which may not be able to be overridden from a template and would always be false when called from a template.

For consistency with other methods accessible from templates, would `Convert` (with capital C` be better?

Reply:
> If we need to pass in options from templates in projects we pass them as json_encoded array or object, and access it using func_get_args as it can't have a default value.

Sorry, I'm confused about what you mean here... this method doesn't exist yet. Where are you passing options to in templates?

> Which brings me to the second param here which may not be able to be overridden from a template and would always be false when called from a template.

You can call `$MyFile.convert('ext', true)` in a template. I have tested this and it works as expected.

> For consistency with other methods accessible from templates, would Convert (with capital C` be better?

The other methods do not follow the current best practices in our [PHP Coding Conventions](https://docs.silverstripe.org/en/5/contributing/coding_conventions/php_coding_conventions/) - but we should adhere to best practices with all new code, I think.

Reply:
> > If we need to pass in options from templates in projects we pass them as json_encoded array or object, and access it using func_get_args as it can't have a default value.
> 
> Sorry, I'm confused about what you mean here... this method doesn't exist outside of this PR... Where are you passing options to in templates?

That was just an example how one can access extra parameters passed in, sorry for the confusion.

> > Which brings me to the second param here which may not be able to be overridden from a template and would always be false when called from a template.
> 
> You can call `$MyFile.convert('ext', true)` in a template. I have tested this and it works as expected.

Would it also work if you did `<% with $MyFile %><a href="$convert('ext', true)"><% end_with %>`?
 
> > For consistency with other methods accessible from templates, would Convert (with capital C` be better?
> 
> The other methods do not follow the current best practices in our [PHP Coding Conventions](https://docs.silverstripe.org/en/5/contributing/coding_conventions/php_coding_conventions/) - but we should adhere to best practices with all new code, I think.

That was my guess right after I asked.

Reply:
> That was just an example how one can access extra parameters passed in, sorry for the confusion.

Ahh okay, fair enough.
Honestly I think the times people will want explicit options will be quite uncommon (options will usually only be supported by one specific converter anyway) - and in those cases there's always the option of handling this in their model or controller.

I personally would prefer to err on the side of a simpler API at this stage, and revisit if we find this is becoming a barrier for people - but I'm happy to change it if the consensus is that we should allow JSON formatted strings from templates here.

> Would it also work if you did `<% with $MyFile %><a href="$convert('ext', true)"><% end_with %>`?

Don't see why not, the syntax is still the same. If that doesn't work it's a bug in the parser which needs to be reported and fixed.

Reply:
I'll try to reproduce the default param not being able to be overridden from a template on a fresh install. Might be just related to global template providers, not to methods on objects like here.

Reply:
Could also be related to getters - `$MyGetter('some param')` won't actually pass the values through to a `getMyGetter()` method, you have to use the full method name in the template like `getMyGetter('some param')`